stellargraph/link_prediction.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from sklearn.pipeline import Pipeline
+from sklearn.linear_model import LogisticRegressionCV
+from sklearn.metrics import roc_auc_score
+from sklearn.preprocessing import StandardScaler
+import numpy as np
+import warnings
+
+
+def _operator_hadamard(u, v):
+    return u * v
+
+
+def _operator_avg(u, v):
+    return (u + v) / 2.0
+
+
+def _operator_l1(u, v):
+    return np.abs(u - v)
+
+
+def _operator_l2(u, v):
+    return (u - v) ** 2
+
+
+_known_operators = {
+    "h": _operator_hadamard,
+    "avg": _operator_avg,
+    "l1": _operator_l1,
+    "l2": _operator_l2,
+}
+
+
+def _get_operator(binary_operator):
+    if callable(binary_operator):
+        # the argument is a callable itself
+        return binary_operator
+    try:
+        return _known_operators[binary_operator]
+    except:
+        # catch all exceptions to handle non-hashability, etc
+        raise ValueError(f"Unexpected binary operator: {binary_operator}")
+
+
+class LinkPredictionClassifier:
+    """
+    Link Prediction Classifier designed to be used with an existing node representation learner.
+
+    Args:
+        transform_node (callable): A function to transform node IDs into node features.
+        binary_operators (list of str, optional): Binary operators applied on node features to
+            produce the corresponding edge feature. Each binary operator will be used to train and
+            predict a separate classifier. By default, all binary operators are used.
+
+    """
+
+    def __init__(self, transform_node, binary_operators=None):
+        self.transform_node = transform_node
+        if binary_operators is not None:
+            self._binary_operators = binary_operators
```

Comment:
This doesn't validate that these are valid; why did you take this approach?

Reply:
Ah hmm sorry I missed that. I did still want to keep the original strings in order to create entries in the other dicts later e.g. `self._classifiers`. I guess this could be
```python
if binary_operators is not None:
    self._binary_operators = {op: _get_operator(op) for op in binary_operators}
else:
    self._binary_operators = _known_operators
```